Mathlib/RingTheory/Smooth/StandardSmooth.lean

```diff
@@ -0,0 +1,208 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jung Tao Cheng, Christian Merten, Andrew Yang
+-/
+import Mathlib.Algebra.MvPolynomial.PDeriv
+import Mathlib.LinearAlgebra.Determinant
+import Mathlib.RingTheory.Presentation
+
+/-!
+# Standard smooth algebras
+
+In this file we define standard smooth presentations for algebras. This notion
+is introduced for `SubmersivePresentations`, which is a `Presentation`
+that has less relations than generators. More precisely there exists
+an injective map from `P.relations` to `P.vars`.
+
+Finally, we introduce the notion of standard smooth algebras, which are algebras that
+admit a standard smooth presentation.
+
+## Main definitions
+
+All of these are in the `Algebra` namespace. Let `S` be an `R`-algebra.
+
+- `SubmersivePresentation`: A `Presentation` of `S` as `R`-algebra, equipped with an injective map
+  from `P.relations` to `P.vars`. This map is used to define the differential of a submersive
+  presentation.
+
+Let now `P` be a submersive presentation of `S` over `R`.
+
+- `SubmersivePresentation.differential`: A linear endomorphism of `P.rels → P.ring` sending
+  the `j`-th standard basis vector, corresponding to the `j`-th relation, to the vector
+  of partial derivatives of `P.relation j` with respect to the coordinates `P.map i` for
+  `i : P.rels`.
+- `SubmersivePresentation.jacobian`: The determinant of `P.differential`.
+- `SubmersivePresentation.jacobiMatrix`: If `P.rels` has a `Fintype` instance, we may form
+  the matrix corresponding to `P.differential`. Its determinant is `P.jacobian`.
+- `SubmersivePresentation.IsStandardSmooth`: `P` is standard smooth, if it's jacobian
+  is a unit in `S` and it is finite.
+
+Furthermore, we define:
+
+- `IsStandardSmooth`: `S` is `R`-standard smooth if `S` has a standard smooth submersive
+  `R`-presentation.
+- `IsOfRelativeDimension n`: A standard smooth `R`-algebra `S` is of relative dimension `n`
+  if `S` has a standard smooth submersive `R`-presentation of dimension `n`. Equivalently,
+  that every standard smooth has dimension `n` (TODO, see below).
+- `IsStandardSmoothOfRelativeDimension n`: The combination of the first two.
+
+Finally, in the `RingHom` namespace we define
+
+- `IsStandardSmooth`: A ring hom `R →+* S` is standard smooth if `S` is standard smooth as
+  `R`-algebra.
+
+## TODO
+
+- Show that the canonical presentation for localization away from an element is standard smooth
+  of relative dimension 0.
+- Show that the base change of a standard smooth presentation is standard smooth of equal relative
+  dimension.
+- Show that the composition of standard smooth presentations of relative dimensions `n` and `m` is
+  standard smooth of relative dimension `n + m`.
+- Show that the Kaehler differentials of a standard smooth `R`-algebra `S` of relative dimension `n`
+  are `S`-free of rank `n`. In particular this shows that the relative dimension is independent
```

Comment:
... the module of ... is ... ? (A Kähler differential cannot be `S`-free itself)